### ISSUE TITLE
TypeError: seq.import is not a function when using Typescript

### ISSUE THREAD

**Issue by blazestudios23** [sequelize/sequelize-auto#412]
When I use type script I get the following error from the auto-generated code:

```
TypeError: seq.import is not a function

// Caused by this code:
function load(filePath: string) {
		return seq.import(path.join(__dirname, filePath));
	}
```

The latest version of Serialize removed the import function, which is causing this issue. 

Below is reported to be a fix:

```
const fs = require('fs')
const path = require('path')
const Sequelize = require('sequelize')
const config = require('../config/config')
const db = {}

const sequelize = new Sequelize(
  config.db.database,
  config.db.user,
  config.db.password,
  config.db.options
)

fs
  .readdirSync(__dirname)
  .filter((file) =>
    file !== 'index.js'
  )
  .forEach((file) => {
    /**
     * Choose any one of the below definition for model variable 
     */
    // for individual model files having `export default (sequelize, DataTypes) => {`    
    const model = require(path.join(__dirname, file)).default(sequelize, Sequelize.DataTypes); 
    // for individual model files having `module.exports = (sequelize, DataTypes) => {`    
    const model = require(path.join(__dirname, file))(sequelize, Sequelize.DataTypes);
    db[model.name] = model;
  })

db.sequelize = sequelize
db.Sequelize = Sequelize

module.exports = db
```

**Comment by steveschmitt** [comment]
Yes, the TypeScript generation needs some fixing.  Thanks for reminding me :)

**Comment by papb** [comment]
The `sequelize.import` function was deprecated for some time and was now removed in `v6`. If it helps, a new suggested way of organizing imports is available at https://github.com/sequelize/express-example

**Comment by blazestudios23** [comment]
Thanks @papb , Will this give me my typescript types? I'm really trying to get types into my API.

**Comment by papb** [comment]
@blazestudios23 Yes, it should! Using `sequelize.import` would not give you types because of the dynamic `require` calls happening under the hood, but if you follow the new express-example, which only uses static imports, you should get your types just fine. Let me know how it goes!

**Comment by blazestudios23** [comment]
This is the example I see:

```
const { Sequelize } = require('sequelize');
const { applyExtraSetup } = require('./extra-setup');

// In a real app, you should keep the database connection URL as an environment variable.
// But for this example, we will just use a local SQLite database.
// const sequelize = new Sequelize(process.env.DB_CONNECTION_URL);
const sequelize = new Sequelize({
	dialect: 'sqlite',
	storage: 'sqlite-example-database/example-db.sqlite',
	logQueryParameters: true,
	benchmark: true
});

const modelDefiners = [
	require('./models/user.model'),
	require('./models/instrument.model'),
	require('./models/orchestra.model'),
	// Add more models here...
	// require('./models/item'),
];

// We define all models according to their files.
for (const modelDefiner of modelDefiners) {
	modelDefiner(sequelize);
}

// We execute any extra setup after the models are defined, such as adding associations.
applyExtraSetup(sequelize);

// We export the sequelize connection instance to be used around our app.
module.exports = sequelize;
```

Would something like this work to loop through all the files:

```
fs
  .readdirSync(__dirname)
  .filter((file) =>
    file !== 'index.js'
  )
  .forEach((file) => {
    /**
     * Choose any one of the below definition for model variable 
     */
    // for individual model files having `export default (sequelize, DataTypes) => {`    
    const model = require(path.join(__dirname, file)).default(sequelize, Sequelize.DataTypes); 
    // for individual model files having `module.exports = (sequelize, DataTypes) => {`    
    const model = require(path.join(__dirname, file))(sequelize, Sequelize.DataTypes);
    db[model.name] = model;
  })
```

**Comment by papb** [comment]
@blazestudios23 That approach would not propagate TypeScript types to the models you're putting in the `db` object, due to the dynamic requires.

**Comment by blazestudios23** [comment]
Well, I need to come up with a method that doesn't require writing out every filename as that will get to be pretty tedious and will have to be done every time we generate new models. I prefer the more automated approach.

**Comment by papb** [comment]
@blazestudios23 I am pretty sure every filename has to be written out in order to the TypeScript types propagate correctly. However, this does not mean that you have to do it by hand. Perhaps `sequelize-auto` can generate this file automatically. So the file will have every filename written out, but it will have been written automatically.

@steveschmitt, do you think this is feasible?

**Comment by steveschmitt** [comment]
Yes, `sequelize-auto` could generate a helper function with the file list pre-populated.

**Comment by blazestudios23** [comment]
I'm also getting the following error in my tb.table.ts file:



```
Conversion of type '{ displayElements: typeof sequelize.Model; displayParameters: typeof sequelize.Model; displayPreset: typeof sequelize.Model; displayPresetParameters: typeof sequelize.Model; queryParameters: typeof sequelize.Model; queryStore: typeof sequelize.Model; searchDefinition: typeof sequelize.Model; }' to type 'ITables' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Types of property 'displayElements' are incompatible.
    Type 'typeof Model' is missing the following properties from type 'displayElementsModel': isNewRecord, where, getDataValue, setDataValue, and 12 more.ts(2352)
```

for:

```
// tslint:disable
import * as path from 'path';
import * as sequelize from 'sequelize';
import * as def from './db';

export interface ITables {
	displayElements: def.displayElementsModel;
}

export const getModels = function(seq: sequelize.Sequelize): ITables {

	function load(filePath: string) {
		return seq.import(path.join(__dirname, filePath));
	}

	return {
		displayElements: load('displayElements'),
	} as ITables;
};
```

**Comment by blazestudios23** [comment]
It says :

`Namespace '"c:/Git/VSTS/StateNetApps/Cleaver/node_modules/sequelize/types/index"' has no exported member 'Instance'.ts(2694)`

for:
`export interface queryParametersInstance extends Sequelize.Instance<queryParametersAttribute>, queryParametersAttribute { }
`
Any ideas on how to get around that one?

**Comment by blazestudios23** [comment]
I got it to work by changing the db.tables.ts file to the following:

```
// tslint:disable
import * as path from 'path';
import * as sequelize from 'sequelize';
import * as def from './db';
import { DataTypes } from 'sequelize'; // added this to the file

export interface ITables {
    displayElements: def.displayElementsModel;
}

export const getModels = function(seq: sequelize.Sequelize): ITables {
    function load(filePath: string) {
        return require(path.join(__dirname, filePath))(seq, DataTypes); //change here
    }

    return {
        displayElements: load('displayElements'),
    } as ITables;
};
```

I then added it together as follows in an index.ts file which I think should be auto-generated as well. 

```
import SequelizeAuto from 'sequelize-auto';
import * as dbTables from './db.tables';

const getSavedSearch = () => {};

const models= new SequelizeAuto(
    process.env.DBNAME,
    process.env.USERNAME,
    process.env.PASSWORD,
    {
        host: process.env.ENDPOINT,
        dialect: 'mysql',
        directory: false, // change to  __dirname to import new tables
        typescript: true,
        path: '',
        logging:
            process.env.LOG_SQL_STATEMENTS === 'true' ? console.log : false,
        camelCase: true, // convert snake_case column names to camelCase field names: user_id -> userId
        camelCaseFileName: true, // file names created for each model use camelCase.js not snake_case.js
        additional: {
            timestamps: false
        }
        // tables: [
        //     'display_elements',
        //     'display_parameters',
        //     'myschema.display_preset'
        // ] // use all tables if omitted
        // //...
    }
);

models.run(function(err) {
    if (err) throw err;
});

```
export default dbTables.getModels(models.sequelize);


There are still a lot of issues with incorrect typing, like findOne() and findAll() not even existing in the data type, which may cause my testing to fail.

**Comment by blazestudios23** [comment]
@papb So I got it all working, but when I try to use my API I get the following error:

`ConnectionManager.getConnection was called after the connection manager was closed!`

I did not have this issue before changing the code per above. Does anyone have any ideas here?

**Comment by blazestudios23** [comment]
Just as a record for anyone else going down this route I set the following to fix the above error:

`closeConnectionAutomatically: false`

Now I'm running into a new error that has me stumped:
`SequelizeDatabaseError: Unknown column 'createdAt' in 'field list'`

I tried the following to no avail:

```
additional: {
                timestamps: true,
                createdAt: true,
                updatedAt: true,
                deletedAt: true
            },
```
Does anyone have a suggestion here?

**Comment by blazestudios23** [comment]
The issues is that Sequilize is adding createdAt and updatedAt into my query which don't exist in my table or data model.

**Comment by papb** [comment]
Hi! Sorry for the delay, I will try to look deeper into this still in this week. 

> ```
> additional: {
>                 timestamps: true,
>                 createdAt: true,
>                 updatedAt: true,
>                 deletedAt: true
>             },
> ```

Where did you try this?

**Comment by blazestudios23** [comment]
So based on what I've read setting timestamp to false below is supposed to fix the issue but doesn't:

```
const dataBase = new SequelizeAuto(
        dbName,
        process.env.USERNAME,
        process.env.PASSWORD,
        {
            host: process.env.ENDPOINT,
            dialect: 'mysql',
            directory: directory || false, // change to  __dirname to import new tables
            typescript: true,
            path: '',
            logging:
                process.env.LOG_SQL_STATEMENTS === 'true' ? console.log : false,
            camelCase, // convert snake_case column names to camelCase field names: user_id -> userId
            camelCaseFileName, // file names created for each model use camelCase.js not snake_case.js
            additional: {
                timestamps: false
            },
            tables,
            closeConnectionAutomatically: false
        }
    );
```

I've also set timestamps to false in the model. These are the locations that are supposed to affect this. I'm eagerly looking forward to a fix.

**Comment by blazestudios23** [comment]
I figured out how to solve the issue:

change additional to 
`define: { timestamps: false }`

**Comment by blazestudios23** [comment]
The issue is the example code needs to be changed to reflect this. 

However I've got a lot of TS errors, that will need to be fixed. All the sequize functions are missing from ITabels, like findOne, findAll, query, hasOne, hasMany etc. This definitely needs to be fixed as the TS version is currently not actually useful as it fails all tests.

**Comment by blazestudios23** [comment]
This is the issue I'm running into:

https://github.com/sequelize/sequelize/issues/11215

**Comment by blazestudios23** [comment]
I found a fix, these is what needs to be changed in the code to get it all working: 

change db.tables:
```

import { DataTypes } from 'sequelize';
export interface ITables {
    actions: sequelize.ModelCtor<def.actionsModel>;
}

export const getModels = function(seq: sequelize.Sequelize): ITables {
    function load(filePath: string) {
        return require(path.join(__dirname, filePath))(seq, DataTypes);
    }
    return {
        actions: load('actions'),
     }
}
```

**Comment by blazestudios23** [comment]
Change db.d.ts:

```
export interface actionsAttribute {
    id: string;
}

export interface actionsModel
    extends Sequelize.Model<actionsAttribute>,
        actionsAttribute {}
```

Change <model>.ts:

```
import { actionsModel } from './db';
module.exports = function(sequelize: sequelize.Sequelize, DataTypes) {
    return sequelize.define<actionsModel>('actions', {
        id: {
            type: DataTypes.string(20),
            allowNull: false
    });
};
```

**Comment by blazestudios23** [comment]
Did the code ever get updated to reflect what I wrote above?

**Comment by steveschmitt** [comment]
With the 0.6.4 release, we have completely changed the way we generate TypeScript files.  Now it is more like the [TypeScript usage](https://sequelize.org/master/manual/typescript.html#usage) example in the Sequelize documentation.

The "db" file is now called `init-models.ts`.  There is a brief example of usage in the [README](https://github.com/sequelize/sequelize-auto/blob/master/README.md#typescript).

Please give it a try and let me know how it goes.

**Comment by steveschmitt** [comment]
Fixed in 0.6.5
